### PR TITLE
Infer thresholds for detect_clearsky

### LINKED ISSUE #1808: Extend detect_clearsky for different data periods
The threshold parameters for `pvlib.clearsky.detect_clearsky` were calibrated using data at 1 minute intervals. Recently, Jordan and Hansen [1] showed how to extend the `detect_clearsky` function for time-averaged data out to 30 minute intervals. The extension requires different data window length and different values for the threshold parameters, but doesn't require any change to the parameter calculations. Interestingly, the threshold values are linear functions of the data interval.

**Describe the solution you'd like**
Implement a helper function for `detect_clearsky` that infers the parameter values from the input data interval.

**Describe alternatives you've considered**
Leaving as-is.

**Additional context**

[1] Jordan, D.C. and C. Hansen, "Clear-sky detection for PV degradation analysis using multiple regression", Renewable Energy, v209, p. 393-400, 2023.

### PR DESCRIPTION
**Pull request by ajonesr**
- [x] Closes #1808 
 - [x] I am familiar with the [contributing guidelines](https://pvlib-python.readthedocs.io/en/latest/contributing.html)
 - [x] Tests added
 - ~~[ ] Updates entries in [`docs/sphinx/source/reference`](https://github.com/pvlib/pvlib-python/blob/main/docs/sphinx/source/reference) for API changes.~~
 - [x] Adds description and name entries in the appropriate "what's new" file in [`docs/sphinx/source/whatsnew`](https://github.com/pvlib/pvlib-python/tree/main/docs/sphinx/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
 - [x] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
 - [x] Pull request is nearly complete and ready for detailed review.
 - [x] Maintainer: Appropriate GitHub Labels (including `remote-data`) and Milestone are assigned to the Pull Request and linked Issue.

### CODE REVIEW

File: pvlib/clearsky.py [hunk 1]
```diff
@@ -644,8 +644,30 @@ def _clear_sample_index(clear_windows, samples_per_window, align, H):
     return clear_samples
 
 
-def detect_clearsky(measured, clearsky, times=None, window_length=10,
-                    mean_diff=75, max_diff=75,
+def _clearsky_get_threshold(sample_interval):
+    """
+    Returns threshold values for kwargs in detect_clearsky
+    """
```
Comment (cwhanse) [inline pvlib/clearsky.py:650]: I'd insert the reference here also

    .. [2] Jordan, D.C. and C. Hansen, "Clear-sky detection for PV
       degradation analysis using multiple regression", Renewable Energy,
       v209, p. 393-400, 2023.

**Comment by cwhanse** [comment]
We should add a test that confirms `_clearsky_get_threshold` is called from `detect_clearsky` when `infer_limits`is True. See the use of `mocker.spy` e.g. [here](https://github.com/pvlib/pvlib-python/blob/cfd6e78235e9eff55308f138d4dfe79317b72299/pvlib/tests/test_iam.py#L257).

**Comment by ajonesr** [comment]
> @ajonesr please remove whatsnew/v0.10.0.rst from this PR. The test failures may be a problem with the new private function; somehow the `alpha` variable is getting set to NaN.

Whoops, sorry about that. I think I've fixed it now?

**Comment by cwhanse** [comment]
@arjones still seeing the test error. You can run the tests locally `pytest test_clearsky.py`

**Comment by adriesse** [comment]
Could we have an issue with a brief high-level description linked to this?

**Comment by cwhanse** [comment]
> Could we have an issue with a brief high-level description linked to this?

#1808

**Review by cwhanse — APPROVED** [review]
Thanks @ajonesr

File: pvlib/clearsky.py [hunk 2]
```diff
@@ -773,6 +809,22 @@ def detect_clearsky(measured, clearsky, times=None, window_length=10,
...
+    # if infer_limits, find threshold values using the sample interval
+    if infer_limits:
+        window_length, mean_diff, max_diff, lower_line_length, \
+            upper_line_length, var_diff, slope_dev = \
+                _clearsky_get_threshold(sample_interval)
+
+        # recalculate sample_interval, samples_per_window using returned
+        # window_length
+        sample_interval, samples_per_window = \
+            tools._get_sample_intervals(times, window_length)
```
Comment (kandersolar) [inline pvlib/clearsky.py:821]:
```suggestion
        # recalculate samples_per_window using returned
        # window_length
        _, samples_per_window = \
            tools._get_sample_intervals(times, window_length)
```
`sample_interval` shouldn't change, right?  (might be able to reflow this comment and code to fit on a single line now)

File: pvlib/clearsky.py [hunk 3]
```diff
@@ -676,6 +706,9 @@ def detect_clearsky(measured, clearsky, times=None, window_length=10,
     times : DatetimeIndex or None, default None.
         Times of measured and clearsky values. If None the index of measured
         will be used.
+    infer_limits : bool, default False
+        If True, does not use passed in kwargs (or defaults), but instead
+        interpolates these values from Table 1 in [2].
```
Comment (kandersolar) [inline pvlib/clearsky.py:711]:
```suggestion
        interpolates these values from Table 1 in [2]_.
```

File: pvlib/clearsky.py [hunk 4]
```diff
@@ -644,8 +644,38 @@ def _clear_sample_index(clear_windows, samples_per_window, align, H):
...
+    References
+    ----------
+    .. [1] Jordan, D.C. and C. Hansen, "Clear-sky detection for PV
+       degradation analysis using multiple regression", Renewable Energy,
+       v209, p. 393-400, 2023.
+    """
+
+    if (sample_interval < 1 or sample_interval > 30):
+        raise ValueError("sample_interval must be larger than 1 and \
+                          smaller than 30")
```
Comment (kandersolar) [inline pvlib/clearsky.py:661]: `sample_interval` is an internal variable in `detect_clearsky`, so the user won't necessarily know what this means.  Maybe reword to something like this?

> `infer_limits=True` can only be used for inputs with time step from 1 to 30 minutes

**Review by kandersolar — COMMENTED** [review]
Thanks @ajonesr!  Well-structured code.  A few suggestions below.

@cwhanse, perhaps this addition should be followed up with some change to `pvanalytics.features.clearsky.reno`.

File: pvlib/clearsky.py (hunk 2 above)
Comment (ajonesr) [reply pvlib/clearsky.py]: The code line was still a little too long, but the comment does fit on one line now

**Review by kandersolar — APPROVED** [review]
LGTM!

**Comment by kandersolar** [comment]
Thanks @ajonesr and congrats on your first merged PR to pvlib!
